### PR TITLE
 Add support for arbitrarily many routes instead of only gateways. 

### LINKED ISSUE #209: RFC: multiple routes
As asked in #207, I'm opening an RFC issue about multiple routes.

I created a naive implementation of routes [here](https://github.com/m-labs/smoltcp/compare/master...ProgVal:naive-routes), based on simply iterating over a managed slice.

Of course, there are probably more efficient ways to do it. Any thoughts?

### PR DESCRIPTION
**Pull request by progval**
Closes #209.

Note: As this PR uses the newly introduced `managed::ManagedMap::range` method, I changed Cargo.toml to use `managed`'s git repo instead of the released version.

### CODE REVIEW

File: src/wire/ipv4.rs [hunk 1]
```diff
@@ -110,8 +110,8 @@ impl fmt::Display for Address {
 /// subnet masking prefix length.
 #[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Default)]
 pub struct Cidr {
-    address:    Address,
-    prefix_len: u8,
+    pub(crate) address:    Address,
+    pub(crate) prefix_len: u8,
```
Comment (whitequark) [inline src/wire/ipv4.rs]: Why is this necessary? There are accessors and constructors for these...

**Comment by whitequark** [comment]
Seems very good at cursory check, and very useful functionality too!
@dlrobertson Can you do a detailed review please?
cc @batonius

File: src/wire/ipv4.rs (hunk 1 above)
Comment (ProgVal) [reply src/wire/ipv4.rs]: I need it in order to write this:

```
const CIDR_1: Ipv4Cidr = Ipv4Cidr {      
    address: Ipv4Address([192, 0, 2, 0]),
    prefix_len: 24,                      
};                                       
```

Using a constructor is forbidden in const expressions.

Comment (whitequark) [reply src/wire/ipv4.rs]: Can you make it a function instead? We can make `new` `const fn` when they land. I feel like making a change this big just for tests isn't worth it.

Comment (ProgVal) [reply src/wire/ipv4.rs]: Sure. Done.

File: src/iface/ethernet.rs [hunk 2]
```diff
@@ -152,43 +144,23 @@ impl<'b, 'c, DeviceT> InterfaceBuilder<'b, 'c, DeviceT>
...
-        self.ipv4_gateway = Some(addr);
-        self
-    }
-
-    /// Set the IPv6 gateway the interface will use. See also
-    /// [ipv6_gateway].
-    ///
-    /// # Panics
-    /// This function panics if the given address is not unicast.
+    /// This function panics if any of the addresses are not unicast.
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: I don't think this is true any more.

File: src/iface/route.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    /// This function panics if any of the addresses are not unicast.
+    pub fn update_routes<F: FnOnce(&mut ManagedMap<'a, IpCidr, Route>)>(&mut self, f: F) {
+        f(&mut self.storage);
+    }
+
+    /// Add a default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    ///
+    /// If not enough storage, returns the gateway address.
```
Comment (dlrobertson) [inline src/iface/route.rs]: Typically `Error::Exhausted` is returned if not enough storage. Is three a reason to return the previous gateway addr on error?

File: src/iface/route.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            Ok(route) => Ok(route),
+            Err((_cidr, route)) => Err(route.via_router)
+        }
+    }
+
+    /// Add a default ipv6 gateway (ie. "ip -6 route add ::/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    ///
+    /// If not enough storage, returns the gateway address.
```
Comment (dlrobertson) [inline src/iface/route.rs]: Same.

File: src/iface/route.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    pub fn new<T>(storage: T) -> Routes<'a>
+            where T: Into<ManagedMap<'a, IpCidr, Route>> {
+        let storage = storage.into();
+        Routes { storage }
+    }
+
+    /// Update the routes of this node.
+    ///
+    /// # Panics
+    /// This function panics if any of the addresses are not unicast.
```
Comment (dlrobertson) [inline src/iface/route.rs]: I don't believe this is true now either.

File: src/iface/route.rs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        let route = Route::new_ipv6_gateway(gateway);
+        match self.storage.insert(cidr, route) {
+            Ok(route) => Ok(route),
+            Err((_cidr, route)) => Err(route.via_router)
+        }
+    }
+
+    pub(crate) fn lookup(&self, addr: &IpAddress, timestamp: Instant) ->
+            Option<IpAddress> {
+        if addr.is_broadcast() {
```
Comment (dlrobertson) [inline src/iface/route.rs]: Probably shouldn't run `lookup` on multicast addresses either. I think this can be `is_unicast`.

File: src/iface/route.rs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            if route.via_router != *addr && prefix.contains_addr(addr) {
+                return Some(route.via_router);
+            }
+        }
+
+        None
+    }
+}
+
+#[cfg(all(test, feature = "proto-ipv4"))]
```
Comment (dlrobertson) [inline src/iface/route.rs]: These tests should run if either `proto-ipv4` or `proto-ipv6` are enabled. See `MOCK_IP_ADDR_*` in [wire/ip.rs:852-876](https://github.com/m-labs/smoltcp/blob/master/src/wire/ip.rs#L853-L876). You will need to add `MOCK_IP_CIDR_*`.

File: src/iface/route.rs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        let cidr = match addr {
+            #[cfg(feature = "proto-ipv4")]
+            IpAddress::Ipv4(addr) => IpCidr::Ipv4(Ipv4Cidr::new(*addr, 32)),
+            #[cfg(feature = "proto-ipv6")]
+            IpAddress::Ipv6(addr) => IpCidr::Ipv6(Ipv6Cidr::new(*addr, 128)),
+            _ => unimplemented!()
+        };
+
+        for (prefix, route) in self.storage.range((Bound::Unbounded, Bound::Included(cidr))).rev() {
+            // TODO: do something with route.preferred_until
```
Comment (dlrobertson) [inline src/iface/route.rs]: You might want to create a FIXME or issue to research a clever way to clear out expired routes.

File: src/iface/route.rs [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            if let Some(expires_at) = route.expires_at {
+                if timestamp > expires_at {
+                    continue;
+                }
+            }
+
+            if route.via_router != *addr && prefix.contains_addr(addr) {
```
Comment (dlrobertson) [inline src/iface/route.rs]: I don't think this is necessary. If the address we're querying is the address of the router we either didn't run address resolution or it didn't respond. If `lookup` returns the address of the router we'll either used the cached link layer address or run address resolution.

File: src/iface/route.rs [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    const ADDR_1A: Ipv4Address = Ipv4Address([192, 0, 2, 1]);
+    const ADDR_1B: Ipv4Address = Ipv4Address([192, 0, 2, 13]);
+    const ADDR_1C: Ipv4Address = Ipv4Address([192, 0, 2, 42]);
+    fn cidr_1() -> Ipv4Cidr {
+        Ipv4Cidr::new(Ipv4Address([192, 0, 2, 0]), 24)
+    }
+
+    const ADDR_2A: Ipv4Address = Ipv4Address([198, 51, 100, 1]);
+    const ADDR_2B: Ipv4Address = Ipv4Address([198, 51, 100, 21]);
+    fn cidr_2() -> Ipv4Cidr {
```
Comment (dlrobertson) [inline src/iface/route.rs]: What is the purpose of this test?

File: src/iface/route.rs [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+/// On systems without heap, use:
+///
+/// ```rust
+/// use smoltcp::iface::Routes;
+/// let mut routes_storage = [];
+/// let mut routes = Routes::new(&mut routes_storage[..]);
+/// ```
+#[derive(Debug)]
+pub struct Routes<'a> {
+    storage:      ManagedMap<'a, IpCidr, Route>,
```
Comment (dlrobertson) [inline src/iface/route.rs]: whitespace

File: src/iface/ethernet.rs (hunk 2 above)
Comment (ProgVal) [reply src/iface/ethernet.rs]: Indeed.

File: src/iface/route.rs (hunk 3 above)
Comment (ProgVal) [reply src/iface/route.rs]: I mimicked ManagedMap::range's return result. I'll change that.

File: src/iface/route.rs (hunk 8 above)
Comment (ProgVal) [reply src/iface/route.rs]: What do you mean by "create a FIXME"?

File: src/iface/route.rs (hunk 7 above)
Comment (ProgVal) [reply src/iface/route.rs]: will do

File: src/iface/route.rs (hunk 10 above)
Comment (ProgVal) [reply src/iface/route.rs]: Which one?

File: src/iface/route.rs (hunk 8 above)
Comment (dlrobertson) [reply src/iface/route.rs]: Add a comment with something like the following.

```
FIXME(ProgVal): Find a clever way to clear out expired routes.
```

I think an issue would be best though.

File: src/iface/ethernet.rs [hunk 12]
```diff
@@ -59,31 +60,25 @@ pub struct Interface<'b, 'c, DeviceT: for<'d> Device<'d>> {
...
-struct InterfaceInner<'b, 'c> {
+struct InterfaceInner<'b, 'c, 'e> {
     neighbor_cache:         NeighborCache<'b>,
     ethernet_addr:          EthernetAddress,
     ip_addrs:               ManagedSlice<'c, IpCidr>,
-    #[cfg(feature = "proto-ipv4")]
-    ipv4_gateway:           Option<Ipv4Address>,
-    #[cfg(feature = "proto-ipv6")]
-    ipv6_gateway:           Option<Ipv6Address>,
+    routes:                 Routes<'e>,
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: Is there a reason to not use an `Option` here? Then you don't have to run `lookup` if the user hasn't configured a buffer for routes.

**Comment by dlrobertson** [comment]
One more small question, but it looks really good!

File: src/iface/ethernet.rs (hunk 12 above)
Comment (ProgVal) [reply src/iface/ethernet.rs]: Not really. Is it really worth it? `lookup` should exit very quickly if there are no routes.

File: src/iface/route.rs (hunk 8 above)
Comment (ProgVal) [reply src/iface/route.rs]: Done: https://github.com/m-labs/smoltcp/issues/221

File: examples/client.rs [hunk 13]
```diff
@@ -40,11 +40,14 @@ fn main() {
     let ethernet_addr = EthernetAddress([0x02, 0x00, 0x00, 0x00, 0x00, 0x02]);
     let ip_addrs = [IpCidr::new(IpAddress::v4(192, 168, 69, 2), 24)];
     let default_v4_gw = Ipv4Address::new(192, 168, 69, 100);
+    let mut routes_storage = [None];
```
Comment (whitequark) [inline examples/client.rs]: Nit: `[None; 1]` here and `[None; 2]` in the other example as a demonstration that you don't have to list every item explicitly.

File: src/iface/route.rs [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+    /// Creates a routing tables. The backing storage is **not** cleared
+    /// upon creation.
+    pub fn new<T>(storage: T) -> Routes<'a>
+            where T: Into<ManagedMap<'a, IpCidr, Route>> {
+        let storage = storage.into();
+        Routes { storage }
+    }
+
+    /// Update the routes of this node.
+    pub fn update_routes<F: FnOnce(&mut ManagedMap<'a, IpCidr, Route>)>(&mut self, f: F) {
```
Comment (whitequark) [inline src/iface/route.rs]: This should just be `update`.

File: src/iface/route.rs [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+        f(&mut self.storage);
+    }
+
+    /// Add a default ipv4 gateway (ie. "ip route add 0.0.0.0/0 via `gateway`").
+    ///
+    /// On success, returns the previous default gateway, if any.
+    #[cfg(feature = "proto-ipv4")]
+    pub fn add_default_ipv4_gateway(&mut self, gateway: Ipv4Address) -> Result<Option<Route>> {
```
Comment (whitequark) [inline src/iface/route.rs]: This adds a default IPv4 *route* (i.e. a single route is added), not *gateway* (which would imply that other routes might use the same gateway if none is specified). Also docstring and IPv6 needs to be changed.

**Comment by ProgVal** [comment]
Any idea why Travis suddenly started failing at building `managed`?

**Comment by whitequark** [comment]
You can bump version of `managed` to `0.7.0`.

**Comment by whitequark** [comment]
Sorry, I didn't notice you updated the PR. It looks good to me now. Can you squash the commits?

**Comment by ProgVal** [comment]
Done!

**Comment by whitequark** [comment]
@m-labs-homu r+

**Comment by m-labs-homu** [comment]
:pushpin: Commit f81d902 has been approved by `whitequark`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit f81d9028ac9befe4cc880a658b51f5a36e50371f with merge 06d18130cd8b25edbcad207bbdbb157f1c1203df...

**Comment by m-labs-homu** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/m-labs/smoltcp/builds/390228449?utm_source=github_status&utm_medium=notification)
Approved by: whitequark
Pushing 06d18130cd8b25edbcad207bbdbb157f1c1203df to master...
